apps/grades-sync/src/job.rs

```diff
@@ -0,0 +1,203 @@
+        futures::future::join_all(map.into_values().map(|grades| {
+                async move {
+                    for grade in grades {
+                        self.synchronize_single_grade(grade).await.unwrap();
+                    }
+                }
+            })).await;
+        info!("synchronization complete");
+        Ok(())
+    }
+
+    async fn synchronize_single_faculty(&self, department: HkdirDepartment) -> anyhow::Result<()> {
+        // Create the faculty if it doesn't exist. The underlying query performs an on
+        // conflict update, which means that we do not need to check if the faculty
+        // exists before creating it.
+        let faculty = self
+            .faculty_repository
+            .create_faculty(department.faculty_code, department.faculty_name)
+            .await
+            .unwrap();
+
+        // Create the department. The same rules apply here as for the faculty.
+        self.department_repository
+            .create_department(
+                department.department_name,
+                department.department_code,
+                faculty.id,
+            )
+            .await
+            .unwrap();
+        Ok(())
+    }
+
+    async fn synchronize_single_subject(&self, subject: HkdirSubject) -> anyhow::Result<()> {
+        // Forge a slug from the subject name. A lot of the subject code fields from
+        // HKDir have a -1 or another number appended to them, which is not useful
+        // because we're using to seeing 'TDT4120' instead of 'TDT4120-1'.
+        let re = Regex::new("-[A-Za-z0-9]+$").unwrap();
+        let slug = re.replace_all(&subject.subject_code, "").to_lowercase();
+        // Find a reference to the department.
+        let department = self
+            .department_repository
+            .get_department_by_ref_id(subject.department_code)
+            .await
+            .unwrap();
+
+        // Create the subject if it doesn't exist. The underlying query performs an on
+        // conflict update, which means that we do not need to check if the subject
+        // exists before creating it.
+        self.subject_repository
+            .create_subject(
+                subject.subject_code,
+                subject.subject_name,
+                department.id,
+                slug,
+                subject.instruction_language,
+                subject.level_code,
+                subject.credits.parse().unwrap(),
+                0f32,
+                0,
+                0,
+            )
+            .await
+            .unwrap();
+        Ok(())
+    }
+
+    async fn synchronize_single_grade(&self, grade: HkdirGrade) -> anyhow::Result<()> {
+        // First we find the matching subject, or else we return early.
+        let subject = match self
+            .subject_repository
+            .find_subject_by_ref_id(grade.subject_code)
+            .await?
+        {
+            Some(subject) => subject,
+            None => return Ok(()),
```

Comment:
Is it really a success if the grade wasnt found? just wondering

Reply:
It means that there's some data inconsistency from the API. Earlier versions have just silent-failed.

I think I'll probably just toss in a warn!() or something